### ISSUE TITLE
`$in` has a confusing name

### ISSUE THREAD

**Issue by amtoine** [nushell/nushell#9551]
mentionned in the [meeting from 2023-06-28](https://hackmd.io/O9zc86C9Q02MAUasluTYgQ?view)

### Related problem

when writing a command that's just a pipe, one can do
```nushell
def foo [] {
    $in | lines | ...
}
```
but we really would like to write
```nushell
def foo [] {
    lines | ...
}
```

so maybe `$in` is a bit to *cute* :open_mouth: 

### Describe the solution you'd like

we might want to rename the `$in` variable:
- `$pipe`
- `$collect`
- `$supercalifragilisticexpialidocious` (cc/ @fdncred :wink:)
- ...

**Comment by hustcer** [comment]
IMO, `in` is better than `pipe` or other names

**Comment by fdncred** [comment]
I posted these `other` ideas in discord. I think my goal would be to be less confusing, but it's hard to be less confusing and still have brevity. For me, I've rather be more verbose and less confusing that to just have some symbol or something super brief that no one really knows what it means.
```
$piped_in
$in_pipe
$from_in
$from_pipe
$inline
$collect_in
$in2val
$in-as-value
```

**Comment by Luis-Licea** [comment]
Is it possible for the `$in` variable name to stay the same, but return a warning whenever it is used before a pipe symbol `|`? For example:
```
> $in | lines | echo $in
Error: nu::shell::expected_pipeline

  × Input type not supported.
   ╭─[entry #1:1:1]
 1 │ $in | lines | echo $in
   · ─┬─
   ·  ╰── $in variable must be used after a pipeline. Remove $in If this is a function that accepts output from a pipeline.
   ╰────
```

**Comment by jntrnr** [comment]
@Luis-Licea ooooh that's an idea

**Comment by hyiltiz** [comment]
I'd prefer verbosity as well, and maybe we should have a verbose name and a short name (following the spirit of `-v` and `--verbose`) so a user can elect to use the shortcut alias `$in` if convenience is desired, or the full name `$value_from_pipe_from` / `$intake_from_pipe` if a more verbose style is desired for maintainable code.

**Comment by bobhy** [comment]
The alternatives listed above don't clarify greatly, but I wasn't too confused about `$in` in the first place.  Are you hearing cries of anguish from the user community?

But, specifically in the context of a defined function, why not allow user to declare *name* as well as type of the input?  I found this proposal in the meeting notes (which, thanks for posting!)
```
def foo [
    x: int,
    y: int,
]: string -> list<string> {
  # ...
}
```
Was this intended to allow user to declare a function whose input is of type `string`?  Syntax could be extended to include a name:

```
def foo [
    x: int,
    y: int,
]: from_pipe: string -> list<string> {
  # ...
}
```

And developer can name pipe input just like any other parameter.  

I'd really like to see defined functions able to declare type of input (and not need input type error checking ever again)!

**Comment by fdncred** [comment]
> The alternatives listed above don't clarify greatly, but I wasn't too confused about $in in the first place. Are you hearing cries of anguish from the user community?

I think the main problem is that people assume `$in` in the input from the pipeline. The name kind of indicates that, but it's really `input collected into a nushell value`. They key here is it is collected, i.e. not streaming. I think that's the main point of confusion.

> Was this intended to allow user to declare a function whose input is of type string?

Yes, that is one of the proposals for defining input and output. If you look at our commands, you'll see there is a function as part of the Signature called `input_output_types()`. This syntax, as well as having multiple input_output_types is supposed to feed that function.

**Comment by bobhy** [comment]
What is the difference between streamed and collected input? I'd like to
think collected input presents all the same bits as streamed, but runs the
risk of exhausting memory and failing before it can complete if the input
really is very large.  In that case, 'in' sounds like a fair name for the
thing and the real problem is the different programming model needed for a
defined function to handle streamed vs collected input.
Both unix pipes and powershell (by default) provide a unified model -- the
main processing code is written as a loop body (pipes) or recurrent event
handler (pwsh) that will be executed potentially multiple times until all
input is processed.  What's the Nushell equivalent of that?
The proposal to restrict the meaning of the input variable to just
collected input makes it sound like Nushell is proceeding in a different,
and perhaps non-shell-normative, direction.

**Comment by fdncred** [comment]
> What is the difference between streamed and collected input? 

Streamed means that it goes through the pipeline one thing at a time. Collected means that all the data is collected into one variable and then passed through the pipeline as a whole thing. Nothing will happen until all the data is collected, so with an infinite stream, nushell would appear to hang because the collection never ends. We typically want to stream things in nushell although there is more streaming work to do.

**Comment by bobhy** [comment]
I agree, Nushell and especially user-written code, should be able to handle
conceptually infinite inputs.  For user-written code, I think that's the
only API there should be, it's what people expect of a shell with a
pipeline operator.  So the question is how to get Nushell to that point
without accumulating a pile of user-written code written to a
soon-to-be-deprecated API in the meantime. And I do think you must
eventually deprecate an API that has the hang failure mode built in -- user
written code should "just work", right?

So that's the risk of renaming `$in` to something like `$pipe-collected`
now -- it creates a non-streaming API and begs the question of what to do
with it in V1.

What work remains to expose full streaming to user-written code?  The
explanation about how streaming works is a bit troubling!  Is it really the
*sender* who decides whether to collect its outputs before putting them
into the pipeline?  If so, the *receiver* cannot unmake that choice, no
matter what API it uses.  Better would be for sender and pipeline to stream
and for receiver to decide whether to collect or stream.  Receiver (alone)
knows what kind of input it is expecting (once the extension for
user-defined functions is implemented to allow receiver to declare type of
input).

**Comment by sholderbach** [comment]
I think we can agree that the naming aspect of this issue has been settled by tradition the `$in` tradeoff in brevity and explicitness. 

Optimizations around collection behavior can happen regardless. (If we don't have an issue that can be opened with a clear scope and survey of the landscape)
